torch/distributed/distributed_c10d.py

```diff
@@ -2779,6 +2780,20 @@ def all_reduce(tensor, op=ReduceOp.SUM, group=None, async_op=False):
         tensor([4.+4.j, 6.+6.j], device='cuda:1') # Rank 1
 
     """
+    # Dynamo has built-in logic to map legacy distributed ops to functional collectives.
+    # Let's redirect to a torch function mode that can mimic this logic outside Dynamo
+    # (e.g., non-strict export implements such a torch function mode).
+    relevant_args = (tensor,)
+    if has_torch_function(relevant_args):
```

Comment:
Does this have performance implications? We would be doing this in eager mode path. Is there a way to turn it off in eager and only turn it on in export?

Reply:
These functions are implemented in C++ and supposed to be super fast, and probably will be dominated by `all_reduce` itself. We'll see if something comes up in tests...

Reply:
I think there is some previous art like: https://www.internalfb.com/code/fbsource/[7b844baf6c235ba6431ed1808eea9302c560f5fd]/xplat/caffe2/torch/nn/parameter.py?lines=92

Reply:
I'm not very comfortable disabling torch function mode, what if there are some that are set up elsewhere to handle inputs or what not...?

Reply:
cc @wconstab just in case you have any opinions. Given that most of our torch.* API's have torch_function handlers, adding them for the `c10d_functional` API's doesn't seem like it should be an overhead killer.